src/adaptors/tcp/tcp_adaptor.c

```diff
@@ -2532,16 +2534,26 @@ static void ADAPTOR_final(void *adaptor_context)
     SET_THREAD_UNKNOWN;
     qd_log(LOG_TCP_ADAPTOR, QD_LOG_INFO, "Shutting down TCP protocol adaptor");
 
-    tcp_context->adaptor_finalizing = true;
-
-    while (DEQ_HEAD(tcp_context->connectors)) {
-        qd_tcp_connector_t *cr   = DEQ_HEAD(tcp_context->connectors);
-        qd_dispatch_delete_tcp_connector(tcp_context->qd, cr);
+    while (DEQ_HEAD(tcp_context->listeners)) {
+        qd_tcp_listener_t *listener   = DEQ_HEAD(tcp_context->listeners);
+        qd_tcp_connection_t *conn = DEQ_HEAD(listener->connections);
+        while (conn) {
+            qd_tcp_connection_t *next_conn = DEQ_NEXT(conn);
+            close_connection_XSIDE_IO(conn);
+            conn = next_conn;
+        }
+        qd_tcp_listener_decref(listener);
```

Comment:
This call to decref may crash if the last closed tcp conn is the last entity holding the listener decref - in that case close_connection_XSIDE_IO will drop the refcount to zero and free the listener.

Same issue applies to the decref of the connector as well.

Fixing that may be tricky...

Without too much "brain smart-making" you could try the following hack:

1. modify the _decref() functions to return "true" if the listener/connector was freed else false
2. at the top of the while loop calll _incref before closing any connections. That ensures that the listener/connector won't be freed by any of the connections closing
3. after closing connections, loop calling _decref() until true is returned.

That's pretty hacky.  You could also fetch the refcount before freeing conns and only call _decref if the total number of conns (DEQ_SIZE) is less than the original refcount.

Hackity hack hack!